server/settings_test.go

```diff
@@ -618,20 +622,78 @@ groups:
 
 					res, err := serverClient.Default.Server.ChangeSettings(&server.ChangeSettingsParams{
 						Body: server.ChangeSettingsBody{
-							AlertManagerURL:   url,
+							AlertManagerURL:   alertManagerURL,
 							AlertManagerRules: rules,
 						},
 						Context: pmmapitests.Context,
 					})
 					require.NoError(t, err)
-					assert.Equal(t, url, res.Payload.Settings.AlertManagerURL)
+					assert.Equal(t, alertManagerURL, res.Payload.Settings.AlertManagerURL)
 					assert.Equal(t, rules, res.Payload.Settings.AlertManagerRules)
 
 					gets, err := serverClient.Default.Server.GetSettings(nil)
 					require.NoError(t, err)
-					assert.Equal(t, url, gets.Payload.Settings.AlertManagerURL)
+					assert.Equal(t, alertManagerURL, gets.Payload.Settings.AlertManagerURL)
 					assert.Equal(t, rules, gets.Payload.Settings.AlertManagerRules)
 
+					t.Run("PrometheusReloaded", func(t *testing.T) {
+						ticker := time.NewTicker(1 * time.Second)
+						timeout := time.NewTimer(10 * time.Second)
+
+						prometheusBaseURL := pmmapitests.BaseURL.ResolveReference(&url.URL{
+							Path: "prometheus/",
+						})
+						client, err := prometheusApi.NewClient(prometheusApi.Config{
+							Address: prometheusBaseURL.String(),
+						})
+						require.NoError(t, err)
+						api := prometheusApiV1.NewAPI(client)
+
+						expected := prometheusApiV1.RuleGroup{
+							Rules: prometheusApiV1.Rules{
+								prometheusApiV1.AlertingRule{
+									Alerts:      []*prometheusApiV1.Alert{},
+									Annotations: model.LabelSet{"summary": "High request latency"},
+									Duration:    600,
+									Health:      "ok",
+									Labels:      model.LabelSet{"severity": "page"},
+									Name:        "HighRequestLatency",
+									Query:       `job:request_latency_seconds:mean5m{job="myjob"} > 0.5`,
+								},
+							},
+							File:     "/srv/prometheus/rules/pmm.rules.yml",
+							Interval: 60,
+							Name:     "example",
+						}
+
+						var found bool
+					L:
+						for { // Prometheus needs some time to reload configs.
+							select {
+							case <-ticker.C:
+								result, err := api.Rules(context.TODO())
```

Comment:
I think the usage of a proper context and early return will reduce the complexity of the current code that uses a combination of the ticker, timeout, select, inner loops, and breaks with labels for what looks like a simple task (as a side note, that combination alone should be a red flag). In pseudocode (not checked):

```
ctx, cancel := context.WithTimeout(context.Background(), 10 * time.Second)
defer cancel()

// make API client

for {
  result, err := api.Rules(ctx)
  require.NoError(t, err) // that could be a ctx timeout

  for _, group := range result.Groups {
    if group.Name == "example" {
      // asserts
      return
    }
  }
}
```